### PR TITLE
Migrate to cats, migrate schema validator to networknt one, remove java.net.URL

### PR DESCRIPTION
**Pull request by asoltysik**
I've split the work because there are a lot of lines changed  - this is the PR number 2, after the bumps.

`ValidatableJsonMethods` (Jackson) and `ValidatableJValue` (json4s) will be gone next PR, as will be `TemporaryJson4sCatsUtils`.

The networknt commit is WIP because the ProcessingMessage format is not there yet.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,22 +28,22 @@ Second working method is `lookupSchema`, receiving Schema key as String or direc
 this method traverse all configured repositories trying to find Schema by its key.
 
 ```scala
-import scalaz.ValidationNel
+import cats.data.ValidatedNel
 import com.fasterxml.jackson.databind.JsonNode
 import com.github.fge.jsonschema.core.report.ProcessingMessage
```
Comment (BenFradet) [inline README.md]: this needs to be changed too, no?

File: src/main/resources/v4metaschema.json [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    "anyOf": { "$ref": "#/definitions/schemaArray" },
+    "oneOf": { "$ref": "#/definitions/schemaArray" },
+    "not": { "$ref": "#" }
+  },
+  "dependencies": {
+    "exclusiveMaximum": [ "maximum" ],
+    "exclusiveMinimum": [ "minimum" ]
+  },
+  "default": {}
+}
```
Comment (BenFradet) [inline src/main/resources/v4metaschema.json]: why is this needed?

File: src/main/scala/com.snowplowanalytics.iglu/client/SchemaCriterion.scala [hunk 3]
```diff
@@ -12,11 +12,10 @@
  */
 package com.snowplowanalytics.iglu.client
 
-// Scalaz
-import scalaz._
-import Scalaz._
+// Cats
+import cats._
+import cats.implicits._
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/SchemaCriterion.scala]: could we have more specific imports?

Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.iglu/client/SchemaCriterion.scala]: this comment is valid throughout the codebase

File: src/main/scala/com.snowplowanalytics.iglu/client/repositories/EmbeddedRepositoryRef.scala [hunk 4]
```diff
@@ -134,21 +130,27 @@ case class EmbeddedRepositoryRef(override val config: RepositoryRefConfig, path:
...
     val schemaPath = s"${path}/schemas/${schemaKey.toPath}"
+    val streamOpt =
+      Option(EmbeddedRepositoryRef.getClass.getResource(schemaPath)).map(_.openStream())
+
     try {
-      JsonLoader.fromResource(schemaPath).some.success
+      streamOpt
+        .map(stream => new ObjectMapper().readTree(stream).asRight[ProcessingMessage]) // TODO: sequence on Validated doesn't work
+        .sequence
+        .toValidated
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/repositories/EmbeddedRepositoryRef.scala]: shouldn't this be a nel?

File: src/main/scala/com.snowplowanalytics.iglu/client/utils/TemporaryJson4sCatsUtils.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package com.snowplowanalytics.iglu.client.utils
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/utils/TemporaryJson4sCatsUtils.scala]: header

File: src/main/scala/com.snowplowanalytics.iglu/client/repositories/HttpRepositoryRef.scala [hunk 6]
```diff
@@ -51,29 +52,31 @@ object HttpRepositoryRef {
...
+   * @param uri the URL to fetch the JSON document from
    * @param apikey optional apikey UUID to aunthenticate in Iglu HTTP repo
    * @return The document at that URL
    */
-  private def getFromUrl(url: URL, apikey: Option[String]): JsonNode = {
-    val connection = url.openConnection()
-    apikey match {
-      case Some(key) => connection.setRequestProperty("apikey", key)
-      case None      => ()
+  private def getFromUri(uri: URI, apikey: Option[String]): Option[JsonNode] = {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/repositories/HttpRepositoryRef.scala]: mmh don't we want a Sync there?

File: src/main/scala/com.snowplowanalytics.iglu/client/utils/TemporaryJson4sCatsUtils.scala [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// Just ignore this in the review, these functions are temporary (will be gone in the next PR)
+object TemporaryJson4sCatsUtils {
+
+  def validatedField[A: Manifest](name: String)(json: JValue)(
+    implicit formats: Formats): ValidatedNel[Throwable, A] =
+    json match {
+      case JObject(fs) =>
+        fs.find(_._1 == name)
+          .map(f => f._2.extract[A].valid)
+          .getOrElse(Validated.invalidNel(new RuntimeException(s"$name: $json")))
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/utils/TemporaryJson4sCatsUtils.scala]: I don't think the throwable brings anything

File: src/main/scala/com.snowplowanalytics.iglu/client/validation/processingMessages.scala [hunk 8]
```diff
@@ -14,32 +14,56 @@ package com.snowplowanalytics.iglu.client
 package validation
 
 // Scala
+
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.iglu/client/validation/processingMessages.scala]: empty line

File: src/test/resources/raw-jsonschema/beer-schema.json [hunk 9]
```diff
@@ -1,5 +1,4 @@
 {
-  "$schema": "http://json-schema.org/schema#",
```
Comment (BenFradet) [inline src/test/resources/raw-jsonschema/beer-schema.json]: why was this removed?

File: src/test/scala/com.snowplowanalytics.iglu.client/ResolverSpec.scala [hunk 10]
```diff
@@ -123,65 +120,66 @@ class ResolverSpec extends Specification with DataTables with ValidationMatchers
...
         "iglu:com.snowplowanalytics.iglu-test/invalid_schema/jsonschema/1-0-0",
         List("Iglu Client Embedded [embedded]", "Iglu Test Embedded [embedded]")
       ),
       "array must have at least one element".toProcessingMessage
+      /*
         .put("domain", "syntax")
         .put("schema", asJsonNode(parse("""{"loadingURI":"#","pointer":""}""")))
         .put("keyword", "required")
-        .toString
+       */ // TODO
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.iglu.client/ResolverSpec.scala]: ?

File: src/test/scala/com.snowplowanalytics.iglu.client/SpecHelpers.scala [hunk 11]
```diff
@@ -74,7 +72,7 @@ object SpecHelpers {
     }
     for (unw <- unwanted) {
       pm.put("unwanted", asJsonNode(unw))
-    }
+    }*/
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.iglu.client/SpecHelpers.scala]: ?

File: src/test/scala/com.snowplowanalytics.iglu.client/validation/RawValidationSpec.scala [hunk 12]
```diff
@@ -82,7 +93,7 @@ class RawValidationSpec extends Specification with DataTables with ValidationMat
                   requiredMissing,
                   None)
                 .toString
-          }
+          }*/
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.iglu.client/validation/RawValidationSpec.scala]: ?

File: src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala [hunk 13]
```diff
@@ -48,58 +49,62 @@ class SelfDescValidationSpec extends Specification with ValidationMatchers {
...
+      )
     )
-  ).map(_.toString)
+    .map(_.toString)
 
-  def e3 = invalidJson.validate(false).leftMap(_.map(_.toString)) must beFailing(invalidExpected)
+  def e3 =
+    invalidJson
+      .validate(false)
+      .leftMap(_.map(_.toString)) must beInvalid // TODO: check expected messages
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala]: ?

File: src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala [hunk 14]
```diff
@@ -48,58 +49,62 @@ class SelfDescValidationSpec extends Specification with ValidationMatchers {
...
-    validJson.validateAndIdentifySchema(true) must beSuccessful(
-      (expectedKey, validJson.get("data")))
+    validJson.validateAndIdentifySchema(true) must beValid((expectedKey, validJson.get("data")))
 
   def e6 =
-    invalidJson.validateAndIdentifySchema(false).leftMap(_.map(_.toString)) must beFailing(
-      invalidExpected)
+    invalidJson
+      .validateAndIdentifySchema(false)
+      .leftMap(_.map(_.toString)) must beInvalid // TODO: Check expected messages
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala]: ?

File: src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala (hunk 13 above)
Comment (asoltysik) [reply src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala]: Just a reminder to test the format after it's decided

File: src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala (hunk 14 above)
Comment (asoltysik) [reply src/test/scala/com.snowplowanalytics.iglu.client/validation/SelfDescValidationSpec.scala]: Just a reminder to test the format after it's decided

File: src/test/resources/raw-jsonschema/beer-schema.json (hunk 9 above)
Comment (asoltysik) [reply src/test/resources/raw-jsonschema/beer-schema.json]: It's a link to json-schema draft v7, I've replaced it with `http://json-schema.org/draft-04/schema#` now.

File: src/main/scala/com.snowplowanalytics.iglu/client/repositories/EmbeddedRepositoryRef.scala (hunk 4 above)
Comment (asoltysik) [reply src/main/scala/com.snowplowanalytics.iglu/client/repositories/EmbeddedRepositoryRef.scala]: Actually @chuwy has a good point that a lot of these `Validated` data types should be just `Eithers`. I would rather change it in #105 though

**Comment by chuwy** [comment]
Closing in favor of https://github.com/snowplow/iglu-scala-client/pull/111
